Maksim panibrat hw1

PanMaksim opened the pull request:
First upload

PanMaksim committed:
Create Maksim_Panibrat_HW1_main.cpp

PanMaksim committed:
Create Maksim_Panibrat_HW1_Vector.hpp

PanMaksim committed:
Create Maksim_Panibrat_HW1_Stack.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Stack.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_main.cpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_main.cpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Stack.hpp

PanMaksim committed:
Create Maksim_Panibrat_HW7_Source.cpp

PanMaksim committed:
Create Maksim_Panibrat_HW7_Functions.h

PanMaksim committed:
Create Maksim_Panibrat_HW7_Guide.cpp

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		}
+		return { *data };
+	}
+	T& back() {
```
может быть проблема висячей ссылки.

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		return { *(data + size - 1) };
+	}
+
+	void print_pair() {
```
не должно быть паблик

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+			std::cout << data[i].first << data[i].second << std::endl;
+		}
+	}
+	void print_arithmetic() {
```
смотри выше замечания

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		}
+	}
+
+	void analize() noexcept {
```
смотри выше замечение

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+class Vector {
+public:
+	Vector() = default;
+	Vector(const size_t& size) {
```
build in type нужно передавать по значение.

MykolaSuperman replied:
```suggestion
	Vector(const size_t& size) : capacity{size}, size{size} {
```

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+	Vector(const size_t& size) {
+		std::cout << __FUNCTION__ << std::endl;
+		capacity = size;
+		data = new ((T*)std::malloc(capacity * sizeof(T)))T[capacity];
+		if (data == nullptr) {
+			throw std::bad_alloc();
+		}
```
```suggestion
                data = (T*)std::malloc(capacity * sizeof(T));
                if (data == nullptr) {
                  throw std::bad_alloc{};
                }
		new (data)T[capacity];
```

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+public:
+	Vector() = default;
+	Vector(const size_t& size) {
+		std::cout << __FUNCTION__ << std::endl;
```
удали дебаг инфо

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		}
+	}
+	Vector(const std::initializer_list<T>& list) : Vector(list.size()) {
+		std::cout << __FUNCTION__ << std::endl;
+		data = new ((T*)std::malloc(list.size() * sizeof(T)))T[list.size()];
+		if (data == nullptr) {
+			throw std::bad_alloc();
+		}
```
уже существует в `Vector(const size_t& size)`

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		for (auto& element : list) {
+			*i = std::move(element);
+			i++;
+			size++;
```
удалить  `size++`

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		}
+		T* i{data};
+		for (auto& element : list) {
+			*i = std::move(element);
```
удалить move

MykolaSuperman replied:
```suggestion
			*i++ = element;
```

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+			*i = std::move(element);
+			i++;
+			size++;
+		}
```
```
if (std::is_trivially_constructible<T>::value || is_fundamental) {
  memcpy()
} else {
  for () {
     = 
  }
}

```

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+	}
+	Vector(T* rhs_begin, T* rhs_end) {
+		std::cout << __FUNCTION__ << std::endl;
+		capacity = std::distance(rhs_begin, rhs_end);
```
capacity == size

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		if (data == nullptr) {
+			throw std::bad_alloc();
+		}
+		for (T* d_begin{ data }; rhs_begin != rhs_end; d_begin++, rhs_begin++) {
```
можно проверить оптимизацию, на придмет побитово копируемых типов

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+				size++;
+		}
+	}
+	Vector(const Vector& rhs) : capacity{ rhs.capacity_() }, size{ rhs.size_() }, data{ new ((T*)std::malloc(capacity * sizeof(T)))T[capacity] } {
```
```suggestion
	Vector(const Vector& rhs) : Vector(rhs.size){
```

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+	}
+	Vector(const Vector& rhs) : capacity{ rhs.capacity_() }, size{ rhs.size_() }, data{ new ((T*)std::malloc(capacity * sizeof(T)))T[capacity] } {
+		std::cout << __FUNCTION__ << std::endl;
+		if constexpr (std::is_arithmetic<T>::value) {
```
constexpr  remove

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		std::cout << __FUNCTION__ << std::endl;
+		if constexpr (std::is_arithmetic<T>::value) {
+			std::memcpy(&data, &rhs.data, size);
+		}
```
else or return

Sam4uk commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+	const T& operator[](const size_t& pos) const {
+		if (pos > size) {
+			throw std::out_of_range ("operator[] out of range");
+		}
```
Я вважаю що перевірка діапазону тут зайва


 UB

Sam4uk commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,443 @@
+		if (size == 0) {
+			return true;
+		}
+		return false;
```
```
return size==0;
```

Sam4uk commented on HomeWork1/Maksim_Panibrat_HW7_Functions.h:
```diff
@@ -0,0 +1,10 @@
+#pragma once
```
А де файл `Maksim_Panibrat_HW7_Functions.cpp`?

Рекомендується реалізовувати функції описані у h - file в cpp -файлі з таким же ім'ям

PanMaksim replied:
Але в мене Guide (який вже перейменував в Function_Caller) знаходиться в окремому файлі, отже мені краще перенести його до Source.cpp (вже Functions_Realizations)?

Sam4uk replied:
[Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#File_Names)
Файл h супроводжує або бібліотеку, або source-file з таким же ім'ям
```
SomeName.h
SomeName.cpp
```
В h ми об'являємо функції(методи, класи , тощо). А в "исходнику" робимо реалізацію. Імена однакові але розширення різні

Sam4uk replied:
@MykolaSolyanko потрібно @PanMaksim  додати до "рев'юверів" моїх branch

Sam4uk reviewed (changes requested):
Спроба робити "рев'ю" чужого коду.

PanMaksim committed:
Update and rename Maksim_Panibrat_HW7_Guide.cpp to Maksim_Panibrat_HW7_Functions_Realizataion.cpp

PanMaksim committed:
Rename Maksim_Panibrat_HW7_Functions.h to Maksim_Panibrat_HW7_Functions_List.h

PanMaksim committed:
Rename Maksim_Panibrat_HW7_Functions_Realizataion.cpp to Maksim_Panibrat_HW7_Function_Caller.cpp

PanMaksim committed:
Update and rename Maksim_Panibrat_HW7_Source.cpp to Maksim_Panibrat_HW7_Functions_Realizations.cpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_main.cpp

PanMaksim commented:
Upload after first reviews

PanMaksim committed:
Update Maksim_Panibrat_HW1_main.cpp

PanMaksim committed:
Update and rename Maksim_Panibrat_HW7_Functions_List.h to Maksim_Panibrat_HW7_Test_Functions_List.h

PanMaksim committed:
Update and rename Maksim_Panibrat_HW7_Functions_Realizations.cpp to Maksim_Panibrat_HW7_Test_Functions_Realizations.cpp

PanMaksim committed:
Update and rename Maksim_Panibrat_HW7_Function_Caller.cpp to Maksim_Panibrat_HW7_Test_Function_Caller.cpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Stack.hpp

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

PanMaksim committed:
Подчистил код (first upload not via browser)

(too long commit comment, my fault)
добавил приватную функцию выделения памяти с проверкой на bad_alloc
добавил проверку на самокопирование у операторов копирования и мува
исправил clear, что бы он вызывал деструктор и освобождал старую память
изменил delete_data, что бы программа не падала при вызове delete_data в функциях после clear
вынес операции внутреннего копирования в отдельную приватную функцию
добавил проверки кидает ли move exception

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

укоротил clear вектора,
исправил параметры функций конструкторов и операторов копирования

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Stack.hpp:
```diff
@@ -0,0 +1,50 @@
+#pragma once
+
```
я думаю что в текущем контетексте не хватает явных инклудов например std::pair. 
Хорошей практикой являеться написание явных инклудов текущего контектса. Не нужно пологаться на косвенное включение через другие файлы

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Stack.hpp:
```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+using pair = std::pair<bool, const char *>;
```
А что дает такой алиас ?
Я думаю что стоит дать более читабельный алиас для типа, нежели просто написать ее сокращенную форму. В вашем контексте это выглядит как просто абстракция без явного описания что это

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		new(data)T[capacity]{};
+	}
+	Vector(const size_t new_size, const T value) : capacity{ new_size }, size{ capacity } {
+		data = alloc_memory(capacity);
+		new(data)T[capacity];
```
```suggestion
		data = alloc_memory(capacity);
		new(data)T[capacity];
```
у вас этот код повторяеться несколько раз, может стоит его вынести в отдельный метод.
Также можно использовать форвардинг конструкторов, у вас уже реализован функционал по выделению памяти в конструкторе `Vector(size_t new_size)`

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+	}
+	Vector(const std::initializer_list<T>& list) : Vector(list.size()) {
+		memory_copy(data, list.begin(), list.size());
+		size = list.size();
```
а зачем тут менять размер если он у вас уже измениться в `Vector(list.size())`

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		memory_copy(data, list.begin(), list.size());
+		size = list.size();
+	}
+	Vector(const T* rhs_begin, const T* rhs_end) {
+		capacity = std::distance(rhs_begin, rhs_end);
+		size = capacity;
+		data = alloc_memory(capacity);
+		new(data)T[capacity];
```
```suggestion
	Vector(const T* rhs_begin, const T* rhs_end) : Vector(std::distance(rhs_begin, rhs_end)) {
```
Постарайтесь в своем коде переиспользовать уже существующий функционал кода.
его очень много

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		}
+	}
+
+	void memory_copy(T* destination, const T* source, const size_t copy_size) {
```
Я думаю что в контексте этого метода  copy_size лишнее?
передавайте такое 
```
memory_copy(const T* source_begin, const T* source_end, T* destination)
```

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		if constexpr (std::is_trivially_constructible<T>::value || std::is_fundamental<T>::value) {
+			std::memcpy(destination, source, copy_size * sizeof(T));
+		}
+		else if constexpr (std::is_nothrow_move_constructible<T>::value) {
```
я думаю вам стоит посмотреть в каком контексте вызываеться этот метод. вы пытаетесь использовать std::move c обьекта параметер которого `const T* source`, т.е. у вас будет просто копирование. Еще раз прочитайте внимательно что такое move семантика, и в каком контексте она применима. Напишите для этого тестируемый код с классом у которого есть перегруженные перемещающие операторы, и посмотрите когда они вызваються

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		return data[pos];
+	}
+
+	size_t capacity_() const noexcept {
```
немного не пойму а зачем нижнее подчеркивание  в имени метода

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		return this->capacity;
+	}
+	size_t size_() const noexcept {
+		return this->size;
```
если у вас была техника именования внутренних переменных класса, например `size_` , то у вас бы небыло проблем с совпадениям имени метода с именем поля класса.

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		if (size == capacity) {
+			capacity *= 2;
+		}
+		T* tmp_arr = alloc_memory(capacity);
```
выделения памяти после каждого добавления элемента в контейнер? а какая же тогда эфективность будет этого метода. Каждый раз идти за памятью при добавления элемнта? Я думаю пользы в этом никакой, и хорошого тоже.
Вы должны переалоцировать память только в том случае если `size == capacity`, а не каждый раз при добавления элемента

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+			capacity *= 2;
+		}
+		T* tmp_arr = alloc_memory(capacity);
+		new(tmp_arr)T[capacity];
```
тоже самое замечание как и для push_front

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+
+	template <typename ... Rest>
+	void emplace_back(Rest&&... rest) {
+		push_back(T{ rest... });
```
у меня есть небольшой вопрос а что делает `std::forward`? какая его функциональность и контекст применения

PanMaksim replied:
forward, в даном случае, изменит какой тип будет передан push_back, в зависимости от типа который получит сам forward (временный или не временный объект).

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+			capacity *= 2;
+		}
+		T* tmp_arr = alloc_memory(capacity);
+		new(tmp_arr)T[capacity];
```
вы не должны каждый раз выделять память при добавления делемента, или вставки

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+	void erase(size_t pos) {
+		pos--;
+		T* tmp_arr = alloc_memory(capacity);
+		new(tmp_arr)T[capacity];
```
как и выше

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+	}
+
+
+	void test_analize() noexcept {
```
этот метод не должен быть частью интерфейса класса. Если вы используете его как для тестирования, то он и должен быть частью вашего тестирования, но не быть частью класса вектор в Pull Request

MykolaSuperman commented on HomeWork1/Maksim_Panibrat_HW1_Vector.hpp:
```diff
@@ -0,0 +1,320 @@
+		}
+		delete_data();
+		data = tmp_arr;
+	}
```
Посмотрите реализацию reserve как у @Sam4uk

MykolaSuperman reviewed (commented):
Для дальнейшего ревью кода прошу сделать две задачи:
1. Разберитесь в именовании файлов(заголовочных и cpp). Вам писал поэтому поводу @Sam4uk . 
2. Перепишите код на то чтобы не выделять память при каждом запросе к вектору. Не должна выделяться память если вы что то добавляете в вектор или удаляете, и .д.. Память должна выделяться только по необходимости

PanMaksim committed:
Переименования и слияние тестовых файлов

Слияние тестовых файлов:
Test_Functions_Realizations и Test_Functions_Caller в Test_Functions.cpp
Переименования переменных класса Vector:
capacity -> capacity_
data -> data_
size -> size_
Переименования функций Vector:
size_() -> size()
capacity_() -> capacity()
Переименования тестовых enum

PanMaksim committed:
Переименование тестовых файлов

PanMaksim committed:
Rename Maksim_Panibrat_HW7_Test_Functions.h to Maksim_Panibrat_HW7_test_functions.h

PanMaksim committed:
Rename Maksim_Panibrat_HW7_Test_Functions.cpp to Maksim_Panibrat_HW7_test_functions.cpp

PanMaksim committed:
Убрал дебаг функции из класов и добавил явные иклуды

PanMaksim committed:
Изменил функцию alloc_memory

Перенес повторение создания массива из выделенной памяти в функцию create_array

PanMaksim committed:
Изменил функцию memory_copy

PanMaksim committed:
Изменил push_back и insert что бы память переалоцировалась не при каждом вызове

Изменил параметры при вызовах memory_copy на более читабельные.
Добавил конструкторам вектора переиспользование конструктора через size.
Изменил дефолтный конструктор без аргументов, что бы он тоже выделял память.
Изменил push_back и insert что бы память переалоцировалась не при каждом вызове.

PanMaksim committed:
Добавил 2 приватные функции и изменил push_front

Добавил приватную функцию вектора  free_space_for_element на сдвиг элементов массива, для последующей вставки нового элемента в освобожденное место.
Добавил приватную функцию ветора place_element на вставку елемента на указанную позицию массива.
Добавил в insert и push_back проверку на возможность move.
Изменил push_front что бы переалокация памяти не происходила при каждом вызове.

PanMaksim committed:
Обновил функции вектора erase и emplace_back

Обновил функции вектора erase, что бы не переалоцировалась память при каждом вызове.
Переименовал функцию delete_data в free_data.
Добавил std::forward в emplace_back.

PanMaksim committed:
Update Maksim_Panibrat_HW1_Vector.hpp

Добавил в reserve проверку new_cap <= capacity_.
Добавил noexcept приватным функциям memory_copy, place_element и free_space_for_element.

PanMaksim committed:
Critical 'reserve' and 'clear' error fix

Исправил критическую ошибку при вызовах reserve, из-за которой при вызове reserve всегда срабатывал new_cap <= capacity_, что всегда вызывало return до выделения новой памяти.
Исправил clear из-за которой в free_data не вызывались деструкторы.

PanMaksim committed:
Added starting and grow factors for capacity_ and fixed clear.

Clear will not deallocating memory.
Default Vector constructor will not allocating memory.
